bleach/html5lib_shim.py

```diff
@@ -37,8 +38,156 @@
 ENTITIES_TRIE = Trie(ENTITIES)
 
 
+class InputStreamWithMemory(object):
+    """Wraps an HTMLInputStream to remember what characters we've seen
+
+    It didn't make sense to implement our own HTMLInputStream, so this
+    wraps the existing ones and keeps track of what we've seen so far. We
+    do this so we can provide the original string the stream had in the case
+    where Bleach's cleaner is going to escape a disallowed tag so we can
+    escape the original string.
+
+    """
+    def __init__(self, inner_stream):
+        self._inner_stream = inner_stream
+        self._buffer = []
+
+    @property
+    def errors(self):
+        return self._inner_stream.errors
+
+    def reset(self):
+        return self._inner_stream.reset()
+
+    def position(self):
+        return self._inner_stream.position()
+
+    def char(self):
+        c = self._inner_stream.char()
+        # char() can return None if EOF, so ignore that
+        if c:
+            self._buffer.append(c)
+        return c
+
+    def charsUntil(self, characters, opposite=False):
+        chars = self._inner_stream.charsUntil(characters, opposite=opposite)
+        self._buffer.extend(list(chars))
+        return chars
+
+    def unget(self, char):
+        if self._buffer:
+            self._buffer.pop(-1)
+        return self._inner_stream.unget(char)
+
+    def stream_history(self):
+        return self._buffer
```

Comment:
This literally just wraps input streams and keeps track of what they've emitted so far. This has the adverse effect of keeping the characters in the stream in memory, but Bleach tends to work on strings that are smallish, so I think this is "ok" for "common use cases".

It's possible we could figure out a way to keep track of just the interesting part of the stream history. Like everything from the most recent `<` that isn't in quotes. That requires some additional finagling and complexity, so I decided not to do that for now.

Reply:
A maxsize might be interesting, but if it fails you end up with missing text and it's unclear why. I think I want to punt on that.

I don't think a deque helps here. The usage of `self._buffer` is in `get_recent_tag_string` where it gets reversed and then turned into a list then the list is iterated over. Using a deque doesn't really help except possibly in the `unget` case but I don't think that triggers often. It's possible that we could rewrite `get_recent_tag_string` in a faster/more-performant fashion. I didn't spend much time thinking about that. If we rewrote it, then a deque might help more.

Reply:
It occurs to me that once `get_recent_tag_string` has been called we know that the stream is a bunch of stuff up to the most recent tag. We'll never want to return anything in that stream history again, so we can dump it. That'll reduce the memory issue.